### PR TITLE
Add updating operating_system in-place feature

### PR DESCRIPTION
**Pull request by TwoDCube**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Relates OR Closes #0000

Output from acceptance testing:
```
% make testacc TEST=./ibm/service/kubernetes/... TESTARGS='-run=TestAccIBMContainerWorkerPoolBasic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./ibm/service/kubernetes/... -v -run=TestAccIBMContainerWorkerPoolBasic -timeout 700m
?       github.com/IBM-Cloud/terraform-provider-ibm/ibm/service/kubernetes/utils/softwaredefinedstorage [no test files]

...

=== RUN   TestAccIBMContainerWorkerPoolBasic
--- PASS: TestAccIBMContainerWorkerPoolBasic (1326.29s)
PASS
ok      github.com/IBM-Cloud/terraform-provider-ibm/ibm/service/kubernetes      1327.785s
```

The VPC acceptance tests did not really make a lot of sense since the update tests are using entirely different region in the base config and the update config so that test is likely broken. Instead, compiled the code and tested manually

Original config:
```
resource "ibm_container_vpc_worker_pool" "worker_pool" {
  flavor           = "bx2.4x16"
  vpc_id           = ibm_is_vpc.vpc.id
  cluster          = ibm_container_vpc_cluster.cluster.id
  worker_count     = 1
  worker_pool_name = "worker-pool"
  operating_system = "UBUNTU_20_64"

  zones {
    name      = "us-south-1"
    subnet_id = ibm_is_subnet.subnet.id
  }
}
```

Changed to 
```
resource "ibm_container_vpc_worker_pool" "worker_pool" {
  flavor           = "bx2.4x16"
  vpc_id           = ibm_is_vpc.vpc.id
  cluster          = ibm_container_vpc_cluster.cluster.id
  worker_count     = 1
  worker_pool_name = "worker-pool"
  operating_system = "UBUNTU_24_64"

  zones {
    name      = "us-south-1"
    subnet_id = ibm_is_subnet.subnet.id
  }
}
```

Output
```
Terraform will perform the following actions:

  # ibm_container_vpc_worker_pool.worker_pool will be updated in-place
  ~ resource "ibm_container_vpc_worker_pool" "worker_pool" {
        id                      = "cr7gt83106uhh817qit0/cr7gt83106uhh817qit0-68490cf"
      ~ operating_system        = "UBUNTU_20_64" -> "UBUNTU_24_64"
        # (11 unchanged attributes hidden)

        # (1 unchanged block hidden)
    }

Plan: 0 to add, 1 to change, 0 to destroy.
ibm_container_vpc_worker_pool.worker_pool: Modifying... [id=cr7gt83106uhh817qit0/cr7gt83106uhh817qit0-68490cf]
ibm_container_vpc_worker_pool.worker_pool: Modifications complete after 2s [id=cr7gt83106uhh817qit0/cr7gt83106uhh817qit0-68490cf]

Apply complete! Resources: 0 added, 1 changed, 0 destroyed.
```

### CODE REVIEW

**Comment by TwoDCube** [comment]
# Testing
Individual worker updates can not be checked in terraform, so I can't make an automated acceptance test to verify if all went good. If anyone has an idea how to do it, feel free to share. Otherwise, here's my manual testing:

## Classic
Initial terraform config
```hcl
resource "ibm_container_cluster" "create_cluster" {
  name               = "zszepesi-tf-1"
  datacenter         = "dal10"
  hardware           = "shared"
  private_vlan_id    = "3424925"
  public_vlan_id     = "3424923"
  machine_type       = "b3c.4x16"
  update_all_workers = false
}

resource "ibm_container_worker_pool" "worker_pool2" {
  cluster          = ibm_container_cluster.create_cluster.id
  machine_type     = "b3c.4x16"
  size_per_zone    = 3
  worker_pool_name = "myworkerpool"
  operating_system = "UBUNTU_20_64"
}
```

Actual operating systems
```
% ibmcloud ks workers -c crm5l8l200elrmm31edg --output json | jq '.[].lifecycle.actualOperatingSystem'
"UBUNTU_20_64"
"UBUNTU_20_64"
"UBUNTU_20_64"
"UBUNTU_20_64"
"UBUNTU_20_64"
```

Desired operating system
```
% ibmcloud ks worker-pool get -c crm5l8l200elrmm31edg -p crm5l8l200elrmm31edg-71bf1e5 --output json | jq '.operatingSystem' 
"UBUNTU_20_64"
```

Made a change to the workerpool and terraform apply
```hcl
resource "ibm_container_worker_pool" "worker_pool2" {
  ...
  operating_system = "UBUNTU_24_64"
}
```

Desired operating system changed
```
% ibmcloud ks worker-pool get -c crm5l8l200elrmm31edg -p crm5l8l200elrmm31edg-71bf1e5 --output json | jq '.operatingSystem'
"UBUNTU_24_64"
```

Ran the `update_all_workers`
```hcl
resource "ibm_container_cluster" "create_cluster" {
  ...
  update_all_workers = true
}
```

All workers in the workerpool got updated
```
% ibmcloud ks workers -c crm5l8l200elrmm31edg --output json | jq '.[].lifecycle.actualOperatingSystem' 
"UBUNTU_20_64" # this is from the default workerpool which did not change OS
"UBUNTU_24_64"
"UBUNTU_24_64"
"UBUNTU_24_64"
"UBUNTU_24_64"
```

## VPC
Initial terraform config
```hcl
resource "ibm_container_vpc_cluster" "cluster" {
  name              = "zszepesi-tf-2"
  kube_version      = "1.30"
  flavor            = "bx2.4x16"
  vpc_id            = ibm_is_vpc.vpc.id
  worker_count      = 1
  resource_group_id = data.ibm_resource_group.resource_group.id
  operating_system  = "UBUNTU_20_64"
  zones {
    subnet_id = ibm_is_subnet.subnet.id
    name      = ibm_is_subnet.subnet.zone
  }
  wait_till = "IngressReady"

  update_all_workers = false
}

resource "ibm_container_vpc_worker_pool" "worker_pool" {
  cluster          = ibm_container_vpc_cluster.cluster.id
  flavor           = "bx2.4x16"
  vpc_id           = ibm_is_vpc.vpc.id
  worker_count     = 4
  worker_pool_name = "worker_pool"
  operating_system = "UBUNTU_20_64"
  zones {
    subnet_id = ibm_is_subnet.subnet.id
    name      = ibm_is_subnet.subnet.zone
  }
}
```

Actual operating systems
```
% ibmcloud ks workers -c crkpch1207u1dvqoiing --output json | jq '.[].lifecycle.actualOperatingSystem'
"UBUNTU_20_64"
"UBUNTU_20_64"
"UBUNTU_20_64"
"UBUNTU_20_64"
"UBUNTU_20_64"
```

Desired operating system
```
% ibmcloud ks worker-pool get -c crkpch1207u1dvqoiing -p crkpch1207u1dvqoiing-643e18c --output json | jq '.operatingSystem'
"UBUNTU_20_64"
```

Made a change to the workerpool and terraform apply
```hcl
resource "ibm_container_vpc_worker_pool" "worker_pool" {
  ...
  operating_system = "UBUNTU_24_64"
}
```

Desired operating system changed
```
% ibmcloud ks worker-pool get -c crkpch1207u1dvqoiing -p crkpch1207u1dvqoiing-643e18c --output json | jq '.operatingSystem'
"UBUNTU_24_64"
```

Ran the `update_all_workers`
```hcl
resource "ibm_container_vpc_cluster" "cluster" {
  ...
  update_all_workers = true
}
```

All workers in the workerpool got updated
```
% ibmcloud ks workers -c crkpch1207u1dvqoiing --output json | jq '.[].lifecycle.actualOperatingSystem' 
"UBUNTU_20_64" # this is from the default workerpool which did not change OS
"UBUNTU_24_64"
"UBUNTU_24_64"
"UBUNTU_24_64"
"UBUNTU_24_64
```

**Comment by jmulsow** [comment]
@hkantare can we get your review so this can make it into the Oct release next week?

**Comment by TwoDCube** [comment]
@hkantare thanks for merging the bluemix-go changes, updated this PR to use that master version and resolved all conflicts

File: ibm/service/kubernetes/resource_ibm_container_cluster.go [hunk 1]
```diff
@@ -980,19 +980,24 @@ func resourceIBMContainerClusterUpdate(d *schema.ResourceData, meta interface{})
...
+				if err != nil {
+					return fmt.Errorf("[ERROR] Error retrieving worker pool: %s", err)
+				}
+
 				/*kubeversion update done if
 				1. There is a change in Major.Minor version
 				2. Therese is a change in patch_version & Traget kube patch version and patch_version are same
 				*/
-				if w.KubeVersion != w.TargetVersion {
+				if w.KubeVersion.Actual != w.KubeVersion.Target || w.LifeCycle.ActualOperatingSystem != workerPool.OperatingSystem {
```
Comment (hkantare) [inline ibm/service/kubernetes/resource_ibm_container_cluster.go:1000]: what is this change for
```
w.LifeCycle.ActualOperatingSystem != workerPool.OperatingSystem 
```

File: ibm/service/kubernetes/resource_ibm_container_vpc_worker_pool.go [hunk 2]
```diff
@@ -515,6 +514,30 @@ func resourceIBMContainerVpcWorkerPoolUpdate(d *schema.ResourceData, meta interf
...
+		Env := v2.ClusterTargetHeader{ResourceGroup: targetEnv.ResourceGroup}
+
+		err = ClusterClient.WorkerPools().SetWorkerPoolOperatingSystem(v2.SetWorkerPoolOperatingSystem{
+			Cluster:         clusterNameOrID,
+			WorkerPool:      workerPoolName,
+			OperatingSystem: operatingSystem,
+		}, Env)
+		if err != nil {
+			return fmt.Errorf("[ERROR] Error updating the operating_system %s: %s", operatingSystem, err)
+		}
```
Comment (hkantare) [inline ibm/service/kubernetes/resource_ibm_container_vpc_worker_pool.go:538]: After we trigger operating system should n't we wait for the operation to be completed and workers comes to normal state.
If we have any dependent operations may be using kubernetes provider will be failed if we don't wait for the update to completed

File: ibm/service/kubernetes/resource_ibm_container_worker_pool.go [hunk 3]
```diff
@@ -432,6 +432,28 @@ func resourceIBMContainerWorkerPoolUpdate(d *schema.ResourceData, meta interface
...
+
+		err = ClusterClient.WorkerPools().SetWorkerPoolOperatingSystem(v2.SetWorkerPoolOperatingSystem{
+			Cluster:         clusterNameorID,
+			WorkerPool:      workerPoolNameorID,
+			OperatingSystem: operatingSystem,
+		}, Env)
+		if err != nil {
+			return fmt.Errorf("[ERROR] Error updating the operating_system %s: %s", operatingSystem, err)
+		}
+	}
```
Comment (hkantare) [inline ibm/service/kubernetes/resource_ibm_container_worker_pool.go:455]: Same as above

File: ibm/service/kubernetes/resource_ibm_container_vpc_worker_pool.go (hunk 2 above)
Comment (TwoDCube) [reply ibm/service/kubernetes/resource_ibm_container_vpc_worker_pool.go:538]: This API call only sets the "desired" operating system and it happens immediately. It does not start the update. The client needs to call the worker update/replace API which will make sure to match the workerpool's desired operating system = actual operating system. The `update_all_workers` will go one by one and do that and wait for the update to finish.

File: ibm/service/kubernetes/resource_ibm_container_cluster.go (hunk 1 above)
Comment (TwoDCube) [reply ibm/service/kubernetes/resource_ibm_container_cluster.go:1000]: So there are 2 cases when you need to call the worker update/replace API, one is if there is a newer BOM version, so the `w.KubeVersion.Actual != w.KubeVersion.Target` and the other is when you set the operating system of the workerpool to something else than what is currently the operating system. This is the code that will do the actual change to the worker's operating system, and here we do wait for the update to complete

File: ibm/service/kubernetes/resource_ibm_container_vpc_worker_pool.go (hunk 2 above)
Comment (jmulsow) [reply ibm/service/kubernetes/resource_ibm_container_vpc_worker_pool.go:538]: @hkantare this is similar to the behavior we have for OS update in the API/CLI/UI. A customer can update the OS property on the worker pool, but then has to roll it out to the worker. We can add some optimizaiton in Terraform for users to opt in to the workers getting updated at the same time the pool does, but that can be added in the next release after this initial release that we need in order to give customers a full 6 months to migrate before before EOS for the exisitng OSes.

**Comment by hkantare** [comment]
Based on https://github.com/IBM-Cloud/terraform-provider-ibm/pull/5589#discussion_r1779841863 link approving PR

**Review by hkantare — APPROVED** [review]

**Pull request merged**